sources/api/bootstrap-containers/src/main.rs

```diff
@@ -0,0 +1,661 @@
+        Some("mark-bootstrap") => Ok((global_args, parse_mark_bootstrap_args(subcommand_args)?)),
+        Some(x) => {
+            return error::Usage {
+                message: format!("Unknown subcommand '{}'", x),
+            }
+            .fail()
+        }
+    }
+}
+
+/// Parses arguments for the 'mark-boostrap' subcommand
+fn parse_mark_bootstrap_args(args: Vec<String>) -> Result<Subcommand> {
+    let mut container_id = None;
+    let mut mode = None;
+
+    let mut iter = args.into_iter();
+    while let Some(arg) = iter.next() {
+        match arg.as_ref() {
+            "--container-id" => {
+                container_id = Some(iter.next().context(error::Usage {
+                    message: "Did not give argument to --container-id",
+                })?);
+            }
+
+            "--mode" => {
+                mode = Some(iter.next().context(error::Usage {
+                    message: "Did not give argument to --mode",
+                })?);
+            }
+
+            x => {
+                return error::Usage {
+                    message: format!("Unexpected argument '{}'", x),
+                }
+                .fail()
+            }
+        }
+    }
+
+    let container_id = container_id.context(error::Usage {
+        message: format!("Did not give argument to --container-id"),
+    })?;
+
+    let mode = mode.context(error::Usage {
+        message: format!("Did not give argument to --mode"),
+    })?;
+
+    Ok(Subcommand::MarkBootstrap(MarkBootstrapArgs {
+        container_id: container_id,
+        // Fail if 'mode' is invalid
+        mode: ValidContainerMode::try_from(mode.to_string())
+            .context(error::BootstrapContainerMode { mode })?,
+    }))
+}
+
+/// Handles how the bootstrap containers' systemd units are created
+fn handle_bootstrap_container<S>(
+    name: S,
+    container_details: &model::BootstrapContainer,
+) -> Result<()>
+where
+    S: AsRef<str>,
+{
+    let name = name.as_ref();
+
+    info!("Handling bootstrap container '{}'", name);
+
+    // Get basic settings, as retrieved from API.
+    let source = container_details
+        .source
+        .as_ref()
+        .context(error::MissingField {
+            name,
+            field: "source",
+        })?;
+
+    let mode = container_details
+        .mode
+        .as_ref()
+        .context(error::MissingField {
+            name,
+            field: "mode",
+        })?;
+
+    // If user data was specified, decode it and write it out
+    if let Some(user_data) = &container_details.user_data {
+        debug!("Decoding user data for container '{}'", name);
+        let decoded_bytes = base64::decode(user_data.as_bytes()).context(error::Base64Decode {
+            base64_string: user_data.as_ref(),
+        })?;
+
+        // Makes sure that the persistent storage directory exists
+        let dir = Path::new(PERSISTENT_STORAGE_DIR).join(name);
+        fs::create_dir_all(&dir).context(error::Mkdir { dir: &dir })?;
```

Comment:
I don't know the answer: is there a potential race if multiple bootstrap containers are trying to create it? Presumably this function doesn't have an issue between time of checking for existence and when it tries to create the dir.

Reply:
I don't think race conditions will occur while creating the directory. `PERSISTENT_STORAGE_DIR` will be created by the `tmpfilesd` configuration files before the bootstrap containers start. And the path `PERSISTENT_STORAGE_DIR/{name}` directory should be unique, because there can't be two bootstrap containers with the same name, since the `bootstrap-containers` settings is a Map.